tools/compression/src/compress.c

```diff
@@ -0,0 +1,283 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+* Included Files
+****************************************************************************/
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <malloc.h>
+#include <stdlib.h>
+#include "../config.h"
+
+#if CONFIG_COMPRESSION_TYPE == 1
+#include "lzma/LzmaLib.h"
+#endif
+
+#define MAX_BLOCK_SIZE 8192
+
+/* Compressio Header */
+struct s_header {
+	int size_header;			/* Size of Compression header */
+	int compression_format;		/* Compression Format */
+	int blocksize;				/* Block Size used for file compression */
+	int sections;				/* No. of compressed blocks for this file */
+	int binary_size;			/* Uncompressed file size */
+	int secoff[];				/* Offsets to Compressed blocks */
+};
+
+enum compression_format {
+	COMPRESSION_NONE = 0,
+	COMPRESSION_LZMA,
+	COMPRESSION_MAX = COMPRESSION_LZMA,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static void show_usage(const char *progname)
+{
+	fprintf(stderr, "USAGE: %s <block size> <compression format> <uncompressed file name> <compressed file name>\n", progname);
+	exit(1);
+}
+
+static void compress(int block_size, int type, char *in_file, char *out_file)
+{
+	unsigned int sections;
+	long unsigned int size;
+	int in_fd;
+	int out_fd;
+	int index;
+	int nbytes;
+	int ret;
+	struct stat buf;
+	unsigned int rpos;
+	long unsigned int readsize = block_size;
+	long unsigned int writesize = block_size;
+	unsigned char *read_buf = NULL;
+	unsigned char *out_buf = NULL;
+	unsigned char *tptr;
+	struct s_header *phdr = NULL;
+	long unsigned int propsSize = LZMA_PROPS_SIZE;
+
+	ret = stat((char *)in_file, &buf);
+	if (ret < 0) {
+		goto error;
```

Comment:
In this case, even though files are not open, it will try to close the files.

Reply:
made changes for this.
files closed only if file descriptor > 0